src/decoding/decodebuffer.rs

```diff
@@ -341,6 +339,7 @@ fn write_all_bytes(mut sink: impl Write, buf: &[u8]) -> (usize, Result<(), Error
     let mut written = 0;
     while written < buf.len() {
         match sink.write(&buf[written..]) {
+            Ok(0) => return (written, Ok(())),
```

Comment:
Here is the crux of the fix. 

In my case I'm reading a single u64, so `sink` can only hold 4 bytes. It's quickly filled, and then subsequent rounds of the `while` loop will write 0 bytes. 


Previously that would lead to an infinite loop.

I'd expect a similar problem for anyone using `collect_to_writer` with a writer with capacity smaller than`buf`.